File: CPP/07 Actions/action_tutorial/src/simple_action_client.cpp
/**
 * @file simple_action_client.cpp
 *
 * @brief A basic ROS2 action client node that sends as goal the number of string concatenation
 *        the action server should perform. 
 *        The server will send back feedbacks and the final result
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */


#include "custom_action/action/concatenate.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"


using Concatenate = custom_action::action::Concatenate;

// The callback for the feedbacks.
// The server sends back the partial sequence of the string concatenation
void FeedbackCallback(
  rclcpp_action::ClientGoalHandle<Concatenate>::SharedPtr,
  const std::shared_ptr<const Concatenate::Feedback> feedback,
  rclcpp::Logger logger)
{
  RCLCPP_INFO(logger,
    "Feedback received: %s",
    feedback->partial_concatenation.c_str());
}


int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto clientNode = rclcpp::Node::make_shared("simple_action_client");
  // action client definition
  // be sure to set the same action name used for the server
  auto actionClient = rclcpp_action::create_client<Concatenate>(clientNode, "concatenation");

  // waiting for the server online response
  if (!actionClient->wait_for_action_server(std::chrono::seconds(10)))
  {
    RCLCPP_ERROR(clientNode->get_logger(), "!!ATTENTION!! Action server not available");
    return 1;
  }

  // creation of the goal message
  auto goalMsg = Concatenate::Goal();
  goalMsg.num_concatenations = 9;  // the number of concatenation to perform

  RCLCPP_INFO(clientNode->get_logger(), "Sending goal");
  // ask server to achieve some goal and wait until it's accepted
  // here some options such as feedbacks are set
  auto sendGoalOptions = rclcpp_action::Client<Concatenate>::SendGoalOptions();
  // binding to make the logger available in the callback
  // otherwise is necessary to declare the node handle as global
  sendGoalOptions.feedback_callback = std::bind(
    FeedbackCallback, std::placeholders::_1, std::placeholders::_2, clientNode->get_logger());
  
  // the goal is sent and the server answers, it's necessary a SUCCESS to continue the communication 
  auto goalHandleFuture = actionClient->async_send_goal(goalMsg, sendGoalOptions);
  if (rclcpp::spin_until_future_complete(clientNode, goalHandleFuture) !=
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_ERROR(clientNode->get_logger(), "send goal call failed :(");
    return 1;
  }

  // the server elaborates the goal and decide to reject or not the goal
  rclcpp_action::ClientGoalHandle<Concatenate>::SharedPtr goalHandle = goalHandleFuture.get();
  if (!goalHandle)
  {
    RCLCPP_ERROR(clientNode->get_logger(), "Goal was rejected by server");
    return 1;
  }

  // wait for the server to be done with the goal
  auto resultFuture = actionClient->async_get_result(goalHandle);
  RCLCPP_INFO(clientNode->get_logger(), "Waiting for result");
  if (rclcpp::spin_until_future_complete(clientNode, resultFuture) !=
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_ERROR(clientNode->get_logger(), "get result call failed :(");
    return 1;
  }

  // the clients gets the final result from the server
  rclcpp_action::ClientGoalHandle<Concatenate>::WrappedResult wrappedResult = resultFuture.get();
  switch (wrappedResult.code)
  {
    case rclcpp_action::ResultCode::SUCCEEDED:
      break;
    case rclcpp_action::ResultCode::ABORTED:
      RCLCPP_ERROR(clientNode->get_logger(), "Goal was aborted");
      return 1;
    case rclcpp_action::ResultCode::CANCELED:
      RCLCPP_ERROR(clientNode->get_logger(), "Goal was canceled");
      return 1;
    default:
      RCLCPP_ERROR(clientNode->get_logger(), "Unknown result code");
      return 1;
  }

  // printout the result
  RCLCPP_INFO(clientNode->get_logger(), "concatenation received");
  RCLCPP_INFO(clientNode->get_logger(), "%s",  wrappedResult.result->final_concatenation.c_str());

  rclcpp::shutdown();
  return 0;
}

File: PYTHON/actions/action_tutorial_py/action_tutorial_py/simple_action_client.py
#  *
#  * @file simple_action_client.py
#  *
#  * @brief A basic ROS2 action client node that sends as goal the number of string concatenation
#  *        the action server should perform. 
#  *        The server will send back feedbacks and the final result
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from action_msgs.msg import GoalStatus
from custom_action.action import Concatenate


def main(args=None):
    rclpy.init(args=args)
    client_node = Node('simple_action_client')
    
    # Action client definition
    # Be sure to set the same action name used for the server
    action_client = ActionClient(client_node, Concatenate, 'concatenation')
    
    # Waiting for the server online response
    if not action_client.wait_for_server(timeout_sec=10.0):
        client_node.get_logger().error('!!ATTENTION!! Action server not available')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    # Creation of the goal message
    goal_msg = Concatenate.Goal()
    goal_msg.num_concatenations = 9  # The number of concatenation to perform
    
    client_node.get_logger().info('Sending goal')
    
    # Create feedback callback as a closure to capture the node
    def feedback_callback(feedback):
        """
        The callback for the feedbacks.
        The server sends back the partial sequence of the string concatenation.
        """
        client_node.get_logger().info(f'Feedback received: {feedback.feedback.partial_concatenation}')
    
    # Ask server to achieve some goal and wait until it's accepted
    # Here some options such as feedbacks are set
    send_goal_future = action_client.send_goal_async(
        goal_msg,
        feedback_callback=feedback_callback
    )
    
    # The goal is sent and the server answers, it's necessary a SUCCESS to continue the communication
    try:
        rclpy.spin_until_future_complete(client_node, send_goal_future)
    except KeyboardInterrupt:
        if rclpy.ok():
            client_node.get_logger().info("Interrupted while waiting for goal response...")
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    goal_handle = send_goal_future.result()
    if not goal_handle:
        client_node.get_logger().error('send goal call failed :(')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    # The server elaborates the goal and decide to reject or not the goal
    if not goal_handle.accepted:
        client_node.get_logger().error('Goal was rejected by server')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    client_node.get_logger().info('Goal accepted by server, waiting for result')
    
    # Wait for the server to be done with the goal
    result_future = goal_handle.get_result_async()
    client_node.get_logger().info('Waiting for result')
    try:
        rclpy.spin_until_future_complete(client_node, result_future)
    except KeyboardInterrupt:
        if rclpy.ok():
            client_node.get_logger().info("Interrupted while waiting for result...")
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    # The clients gets the final result from the server
    wrapped_result = result_future.result()
    if wrapped_result.status == GoalStatus.STATUS_SUCCEEDED:
        client_node.get_logger().info('concatenation received')
        client_node.get_logger().info(f'{wrapped_result.result.final_concatenation}')
    elif wrapped_result.status == GoalStatus.STATUS_ABORTED:
        client_node.get_logger().error('Goal was aborted')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    elif wrapped_result.status == GoalStatus.STATUS_CANCELED:
        client_node.get_logger().error('Goal was canceled')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    else:
        client_node.get_logger().error('Unknown result code')
        client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        return 1
    
    client_node.destroy_node()
    if rclpy.ok():
        rclpy.shutdown()
    return 0


if __name__ == '__main__':
    main()

